internal/goro/adaptive_pool.go

```diff
@@ -0,0 +1,135 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package goro
+
+import (
+	"math/rand"
+	"sync/atomic"
+	"time"
+
+	"go.temporal.io/server/common/clock"
+)
+
+type (
+	// AdaptivePool manages a pool of goroutines to handle small items of work.
+	// The size of the pool starts with minWorkers but can grow if needed, up to maxWorkers,
+	// and can shrink back down to minWorkers.
+	AdaptivePool struct {
+		ts           clock.TimeSource
+		minWorkers   int
+		maxWorkers   int
+		targetDelay  time.Duration
+		shrinkFactor float64
+
+		ch      chan func()
+		stopCh  chan struct{}
+		workers atomic.Int64
+	}
+)
+
+func NewAdaptivePool(
+	ts clock.TimeSource,
+	minWorkers int,
+	maxWorkers int,
+	targetDelay time.Duration,
+	shrinkFactor float64,
+) *AdaptivePool {
+	p := &AdaptivePool{
+		ts:           ts,
+		minWorkers:   minWorkers,
+		maxWorkers:   maxWorkers,
+		targetDelay:  targetDelay,
+		shrinkFactor: shrinkFactor,
+		ch:           make(chan func()),
+		stopCh:       make(chan struct{}),
+	}
+	for i := 0; i < minWorkers; i++ {
+		go p.work()
+	}
+	p.workers.Store(int64(minWorkers))
+	return p
+}
+
+// Stops workers. Note that this does not wait for workers to exit.
+func (p *AdaptivePool) Stop() {
+	close(p.stopCh)
+}
+
+// Do calls f() on a worker. If the call can't be started within targetDelay, it adds another
+// worker.
+func (p *AdaptivePool) Do(f func()) {
+	have := p.workers.Load()
+	if have < int64(p.maxWorkers) {
+		// we might want to add a worker, send with timeout
+		timech, timer := p.ts.NewTimer(p.targetDelay)
+		select {
+		case <-p.stopCh:
+			return
+		case p.ch <- f:
+			timer.Stop()
+			return
+		case <-timech:
+		}
+
+		if p.workers.CompareAndSwap(have, have+1) {
+			go p.work()
+		}
+	}
+
+	// blocking send
+	select {
+	case p.ch <- f:
+	case <-p.stopCh:
+	}
+}
+
+func (p *AdaptivePool) work() {
+	for {
+		have := p.workers.Load()
+		if have > int64(p.minWorkers) {
+			// we might want to exit, receive with timeout
+			// jitter this so we shrink slower than we grow
+			timech, timer := p.ts.NewTimer(time.Duration(float64(p.targetDelay) * p.shrinkFactor * rand.Float64()))
+			select {
+			case <-p.stopCh:
+				return
+			case f := <-p.ch:
+				timer.Stop()
+				f()
+				continue
+			case <-timech:
+			}
+			if p.workers.CompareAndSwap(have, have-1) {
+				return
+			}
+		}
+
+		// blocking receive
+		select {
+		case <-p.stopCh:
+			return
+		case f := <-p.ch:
+			f()
+		}
```

Comment:
Seems like there's no guarantee that all submitted items will be processed before the pool is completely stopped with this approach.

Reply:
That's true, I'll add a comment.

In the server, this won't be stopped unless the whole server is stopping, so there's no point delivering more updates.